src/rustllvm/DemoteSimd.cpp

```diff
@@ -0,0 +1,186 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#include <vector>
+#include <set>
+
+#include "rustllvm.h"
+
+#include "llvm/IR/CallSite.h"
+#include "llvm/IR/Module.h"
+
+using namespace llvm;
+
+static std::vector<Function*>
+GetFunctionsWithSimdArgs(Module *M) {
+  std::vector<Function*> Ret;
+
+  for (auto &F : M->functions()) {
+    // Skip all intrinsic calls as these are always tightly controlled to "work
+    // correctly", so no need to fixup any of these.
+    if (F.isIntrinsic())
+      continue;
+
+    // We know that we started out with a `Module` that has what we want, so
+    // we're just trying to undo specifically the work of the
+    // `ArgumentPromotion` pass. That only runs in a select few circumstances,
+    // so make sure that we don't get anything surprising. For example, make
+    // sure we don't actually return a vector type because rustc shouldn't ever
+    // generate this and nor should passes make this happen.
+    assert(!F->getReturnType()->isVectorTy());
+
+    // If any argument to this function is a by-value vector type, then that's
+    // bad! The compiler didn't generate any functions that looked like this,
+    // and we try to rely on LLVM to not do this! Argument promotion may,
+    // however, promote arguments from behind references. In any case, figure
+    // out if we're interested in demoting this argument.
+    bool anyVector = false;
+    for (auto &Arg : F.args())
+      anyVector = anyVector || Arg.getType()->isVectorTy();
```

Comment:
Shouldn't this stop at the first argument that's a vector?:

```c++
for (auto &Arg : F.args()) {
  if (Arg.getType()->isVectorTy()) {
    Ret.push_back(&F);
    break;
  }
}
```

Reply:
Can one use `llvm/ADT/STLExtras.h` here? This:

```c++
any_of(F.args(), [](auto &arg) { return arg.getType()->isVectorTy().; })
```

might work too.

Reply:
Ah, sure, that works even better.